packages/grid/x-data-grid/src/hooks/utils/useGridSelector.ts

```diff
@@ -1,32 +1,76 @@
 import * as React from 'react';
 import { GridApiCommon } from '../../models/api/gridApiCommon';
 import { OutputSelector } from '../../utils/createSelector';
+import { useLazyRef } from './useLazyRef';
+import { useOnMount } from './useOnMount';
 import { buildWarning } from '../../utils/warning';
+import { fastObjectShallowCompare } from '../../utils/fastObjectShallowCompare';
+
+const stateNotInitializedWarning = buildWarning([
+  'MUI: `useGridSelector` has been called before the initialization of the state.',
+  'This hook can only be used inside the context of the grid.',
+]);
 
 function isOutputSelector<Api extends GridApiCommon, T>(
   selector: any,
 ): selector is OutputSelector<Api['state'], T> {
   return selector.acceptsApiRef;
 }
 
-const stateNotInitializedWarning = buildWarning([
-  'MUI: `useGridSelector` has been called before the initialization of the state.',
-  'This hook can only be used inside the context of the grid.',
-]);
+function applySelector<Api extends GridApiCommon, T>(
+  apiRef: React.MutableRefObject<Api>,
+  selector: ((state: Api['state']) => T) | OutputSelector<Api['state'], T>,
+) {
+  if (isOutputSelector<Api, T>(selector)) {
+    return selector(apiRef);
+  }
+  return selector(apiRef.current.state);
+}
+
+const defaultCompare = Object.is;
+export const shallowCompare = fastObjectShallowCompare;
```

Comment:
Rename `shallowCompare` to `objectShallowCompare`? The reason is the same as for `fastObjectShallowCompare` - to emphasize that it should be used only for objects